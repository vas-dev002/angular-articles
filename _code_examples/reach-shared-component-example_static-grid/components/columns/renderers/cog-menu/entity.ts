import { ICellRendererParams } from "@ag-grid-community/core";

export interface ICogMenuEvent {
	params: ICellRendererParams;
	rowIndex: number;
}
export interface ICogMenuItem {
	css?: string;
	hide?: (event: ICogMenuEvent) => boolean;
	name: string;
	onClick: (event: ICogMenuEvent) => void;
}
export interface ICellCogMenuParams extends ICellRendererParams {
	menuItems: ICogMenuItem[];
	isDisabled?: (cellData: unknown) => boolean;
	getCogMenuTooltip?: (cellData: unknown) => string;
}
