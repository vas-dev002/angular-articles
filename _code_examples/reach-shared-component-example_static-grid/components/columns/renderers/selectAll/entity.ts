import { IHeaderParams } from "@ag-grid-community/core";
import { EventEmitter } from "@angular/core";

export interface ISelectAllMethods {
	selectAll: () => void;
	deselectAll: () => void;
	getRowsCount: (excludeDisabledRows: boolean) => number;
	getSelectedCount: () => number;
	onSelectionChanged: EventEmitter<Object[]>;
}

export interface ISelectAllParams extends IHeaderParams {
	suppressHeaderCheckBox?: boolean;
	selectMethods: ISelectAllMethods;
}
