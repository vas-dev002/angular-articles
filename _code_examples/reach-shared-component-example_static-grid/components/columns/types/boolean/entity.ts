export interface BooleanGroupParams {
	value: string;
}
