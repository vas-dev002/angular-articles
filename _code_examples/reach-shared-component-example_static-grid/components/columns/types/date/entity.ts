import { ValueFormatterParams } from "@ag-grid-community/core";

export interface IDateValue {
	date: string;
}

export interface IDateGroupKeyParams {
	value: string;
	date: string;
}

export interface IDateValueFormatterParams extends ValueFormatterParams {
	value: string | IDateValue;
}
