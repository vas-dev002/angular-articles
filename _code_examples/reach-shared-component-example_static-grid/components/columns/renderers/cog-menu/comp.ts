import { Component } from "@angular/core";
import { ICellRendererAngularComp } from "@ag-grid-community/angular";
import { RowNode } from "@ag-grid-community/core";
import { ICellCogMenuParams, ICogMenuItem } from "./entity";

@Component({
	selector: "child-cell",
	templateUrl: "tpl.html"
})
export class CogMenuSGComponent implements ICellRendererAngularComp {
	public permissionStatus: boolean;
	public params: ICellCogMenuParams;
	public items: ICogMenuItem[];
	public isOpened = false;
	private _getIndex(): number {
		let index = this.params.rowIndex;
		this.params.api.forEachNodeAfterFilterAndSort((rowNode: RowNode, rowIndex: number) => {
			if (this.params.node.id === rowNode.id) {
				index = rowIndex;
			}
		});
		return index;
	}

	public agInit(params: ICellCogMenuParams): void {
		this.params = params;
		this.items = params.menuItems;
	}

	public refresh(): boolean {
		return false;
	}

	public show(item: ICogMenuItem) {
		return item.hide && item.hide({ params: this.params, rowIndex: this._getIndex() });
	}

	public onClick(item: ICogMenuItem) {
		item.onClick({ params: this.params, rowIndex: this._getIndex() });
	}

	public get isDisabled(): boolean {
		return !!this.params && !!this.params.isDisabled && this.params.isDisabled(this.params.data);
	}
}
