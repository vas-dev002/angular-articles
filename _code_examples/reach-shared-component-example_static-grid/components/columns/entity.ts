import { Type, Injector } from "@angular/core";
import { ICellRendererAngularComp } from "@ag-grid-community/angular";
import { ColDef } from "@ag-grid-community/core";
import { IColumnMetaSG } from "../main/entity";
import { StaticGridComponent } from "../main/comp";
import { ColumnDisplayType } from "./types";
import { IColDefAbstractFactory } from "./types/base/base-abstract-factory/models";
import { ObjectDetailsType } from "@evergreen-shared/entity";

export interface ITypeControllerSG {
	renderComponent: Type<ICellRendererAngularComp>;
	columnDefinition(): ColDef;
}

export interface IFilterOptionsSG {
	text: string;
	value: unknown;
	props?: Object;
	sortOrder?: number;
}

export interface ITypeControllerParams {
	meta: IColumnMetaSG;
	grid: StaticGridComponent;
	customOptions: Object;
	injector?: Injector;
}

export interface ILinkMeta {
	uRLColumnName: string;
	objectType: string;
	objectTypeColumnName: string;
	labelColumnName: string;
	keyColumnName: string;
	parameters: string;
	urlParameters: string[];
	renderLinkProperty: string;
}

export interface IOutcomeComparatorValue {
	status: number;
	text: string;
	tooltip: string;
}

export enum ApplicationForwardPathTypeEnum {
	None = 0,
	Retire = 1,
	Keep = 2,
	ForwardPath = 3,
	Uncategorized = 4
}

export enum ApplicationAssociationTypeEnum {
	None = 0,
	Dynamic = 1,
	StickyInclude = 2,
	StickyExclude = 3,
	ForwardPathed = 4,
	Mixed = 5
}

export enum StringValueGetterMode {
	String,
	SelectedItem,
	Params
}

export interface ITypedController {
	key: ColumnDisplayType;
	factory: Type<IColDefAbstractFactory>;
}

export enum TaskFlags {
	IsBucketTask = "IsBucketTask",
	IsReadOnlyTask = "IsReadOnlyTask",
	IsSystemTask = "IsSystemTask"
}

export interface IDateTimeEditorChanges {
	taskId: number;
	projectId: number;
	id: number;
	objectType: ObjectDetailsType;
	date: string;
	postBackParameters: { [key: string]: unknown };
}

export interface ICapacityDateTimeEditorChanges extends IDateTimeEditorChanges {
	slotId: number;
	slotName: string;
}
