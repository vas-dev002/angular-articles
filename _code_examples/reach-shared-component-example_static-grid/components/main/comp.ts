import {
	Component, OnDestroy, ChangeDetectionStrategy, Input, HostListener, EventEmitter, Output, Injector
} from "@angular/core";
import {
	Module,
	GridApi,
	GridOptions,
	ColDef,
	ColumnApi,
	CsvExportParams,
	ColumnEvent,
	RowDragEvent,
	Column,
	ProcessCellForExportParams
} from "@ag-grid-community/core";
import { MenuModule } from "@ag-grid-enterprise/menu";
import { ClientSideRowModelModule } from "@ag-grid-community/client-side-row-model";
import { ServerSideRowModelModule } from "@ag-grid-enterprise/server-side-row-model";
import { ClipboardModule } from "@ag-grid-enterprise/clipboard";
import { CsvExportModule } from "@ag-grid-community/csv-export";
import { RowGroupingModule } from "@ag-grid-enterprise/row-grouping";
import { ColumnsToolPanelModule } from "@ag-grid-enterprise/column-tool-panel";
import {
	getMainDef, frameworkComponents,
	getEditDef, getServiceDef, getSelectionDef,
	getValueByType, isRowSelectable, getDataKey, getDragDef
} from "./config/main";
import {
	IColumnMetaSG,
	IOptionsSG,
	IInternalStateSG,
	IDataRowSG,
	defaultDataKeySG,
	IDragAndDropModel,
	GridMode,
	IOptionsDynamicSG
} from "./entity";
import { merge, intersection } from "lodash";
import { Subject, BehaviorSubject, Observable } from "rxjs";
import { ISelectAllMethods } from "../columns";
import { ColumnControllerSGDirective } from "./directives/column.directive";
import { FilterControllerSGDirective } from "./directives/filter.directive";
import { SortControllerSGDirective } from "./directives/sort.directive";
import { SelectControllerSGDirective } from "./directives/select.directive";
import * as moment from "moment";
import { checkClass } from "../../../../../util/common/element";
import { aggridMenuFix } from "../../../../../util/common/aggridMenuFix";
import { StaticGridApiService } from "./services/static-grid-api.service";
import { getGridSortModel } from "../../../util/fn";
import { tableModalWindowFix } from "../../../../../util/common/style";

@Component({
	selector: "static-grid",
	templateUrl: "./tpl.html",
	changeDetection: ChangeDetectionStrategy.OnPush,
	providers: [StaticGridApiService]
})
export class StaticGridComponent implements OnDestroy {

	// Options
	@Input() public set options(value: IOptionsSG) {
		this._options = value;
		this.setupGrid();
	}
	public get options() {
		return this._options;
	}
	private _options: IOptionsSG;

	@Input() public set dynamicModeOptions(value: IOptionsDynamicSG) {
		this._dynamicModeOptions = value;
		this.setupGrid();
	}
	public get dynamicModeOptions() {
		return this._dynamicModeOptions;
	}
	private _dynamicModeOptions: IOptionsDynamicSG;

	@Input() public set defaultConfig(val: GridOptions) {
		this._defaultConfig = val;
		this.setupGrid();
	}
	public get defaultConfig(): GridOptions {
		return this._defaultConfig;
	}
	private _defaultConfig: GridOptions;

	@Input() public gridMode: GridMode = GridMode.StaticMode;
	@Input() public rowData: IDataRowSG[];
	@Input() public hasData: boolean;
	@Input() public disabled = false;

	public columnController: ColumnControllerSGDirective;
	public filterController: FilterControllerSGDirective;
	public sortController: SortControllerSGDirective;
	public selectController: SelectControllerSGDirective;
	public readonly dateFormat = "ll";
	public readonly dateTimeFormat = "lll";
	public gridStyles: any = {};

	// MetaData
	@Input() public set columnMetaData(value: IColumnMetaSG[]) {
		this._columnMetaData = value;
		this.setupGrid();
	}
	public get columnMetaData() {
		return this._columnMetaData;
	}
	private _columnMetaData: IColumnMetaSG[] = [];


	@Output() visibleColumnsChange = new EventEmitter<ColDef[]>();
	@Output() gridReady = new EventEmitter<StaticGridComponent>();
	@Output() dragAndDropUpdate = new EventEmitter<IDragAndDropModel>();
	@Output() cellUpdate = this.staticGridApiService.cellUpdate;

	public destroy$ = new Subject();
	public columnDefs: ColDef[] = [];
	public gridOptions: GridOptions = {};
	public gridApi: GridApi;
	public gridColumnApi: ColumnApi;

	public agGridModules: Module[] = [
		ClipboardModule, ClientSideRowModelModule,
		CsvExportModule, MenuModule,
		RowGroupingModule, ColumnsToolPanelModule, ServerSideRowModelModule
	];

	public dataKey = defaultDataKeySG;

	// Temporary solution. Used to get data in a directive when referring to the component as a class.
	private readonly _firstDataRendered$: BehaviorSubject<boolean> = new BehaviorSubject(null);
	public firstDataRendered$: Observable<boolean> = this._firstDataRendered$.asObservable();

	constructor(private readonly injector: Injector,
		private readonly staticGridApiService: StaticGridApiService) {}

	private setupGrid() {
		if(!this.options || !this.columnMetaData) {
			return;
		}
		this.configureGridOptions();
		this.configureColDef();
	}

	private configureGridOptions(): void {
		this.gridOptions.frameworkComponents = frameworkComponents;
		if (this.options.enableRowDragAndDrop) {
			const dragDropUpdate = (event: RowDragEvent) => {
				const key = event.node.data[this.dataKey];
				const index = event.node.rowIndex + 1;
				this.dragAndDropUpdate.emit({key: key, index: index});
			};
			this.gridOptions.onRowDragEnd = dragDropUpdate;

			// enable managed drag
			this.gridOptions.rowDragManaged = true;
			this.gridOptions.animateRows = true;
		}
		merge(
			this.gridOptions,
			this.defaultConfig,
			this.options.agGridOptions as GridOptions,
			this.dynamicModeOptions?.agGridOptions as GridOptions
		);
	}

	private getRowCount(excludeDisabledRows = false) {
		if (this.rowData) {
			return (!excludeDisabledRows)
				? this.rowData.length
				: this.rowData.filter((row) =>
					isRowSelectable(row, this.columnMetaData)
				).length;
		} else {
			return 0;
		}
	}

	private getSelectedCount() {
		return this.gridApi ? this.gridApi.getSelectedNodes().length : 0;
	}

	private configureColDef() {
		if (this.columnMetaData && this.columnMetaData.length > 0) {
			this.columnDefs = this.columnMetaData.map(x =>
				merge(
					getMainDef(x, this, this.options.columnControllerOptions, this.injector),
					this.options.editorsByKey
						&& this.options.editorsByKey[x.columnName] ? getEditDef(x, this.options.editorsByKey[x.columnName]) : {}
				)
			);

			if (this.options.serviceColumns) {
				this.options.serviceColumns.forEach((x) => {
					const def = getServiceDef(x);
					const index = (x.index > this.columnDefs.length) ? 0 : x.index;
					this.columnDefs.splice(index, 0, def);
				});
			}

			if (this.options.enableCheckBox) {
				const selectMethods: ISelectAllMethods = {
					getRowsCount: (excludeDisabledRows) => this.getRowCount(excludeDisabledRows),
					getSelectedCount: () => this.getSelectedCount(),
					selectAll: () => this.selectController.selectAllRows(),
					deselectAll: () => this.selectController.deselectAllRows(),
					onSelectionChanged: this.selectController.selectedDataRowsChange
				};

				const showingSelectionOptions = {
					showCheckBoxes: this.selectController.showCheckboxes,
					suppressHeaderCheckBox: this.options.suppressHeaderCheckBox
				};

				this.columnDefs.splice(0, 0,
					getSelectionDef(showingSelectionOptions, selectMethods, this.columnMetaData)
				);
			}

			this.rowDragColDef();
		}
	}

	private rowDragColDef() {
		if (this.options.enableRowDragAndDrop) {
			this.columnDefs.splice(0, 0, getDragDef(this.columnMetaData));
		}
	}

	onColumnVisibilityChanged(event: ColumnEvent) {
		this.columnController.updateGroupColumns(this.gridColumnApi.getRowGroupColumns().map((x) => x.getColId()));
		this.columnController.updateColumnState(this.gridColumnApi.getColumnState());

		const visibleColumns = this.columnController.getVisibleColumns();
		this.visibleColumnsChange.emit(visibleColumns.map(x => x.getColDef()));
	}

	public onReady(params) {
		this.dataKey = getDataKey(this.options, this.columnMetaData);
		this.gridApi = params.api;
		this.gridColumnApi = params.columnApi;
		this.columnController.init(this.gridOptions, this.gridApi, this.gridColumnApi, this.gridMode);
		this.filterController?.init(this.gridOptions, this.gridApi, this.gridColumnApi, this.gridMode);
		this.sortController?.init(this.gridOptions, this.gridApi, this.gridColumnApi, this.gridMode);
		this.selectController?.init(this.gridOptions, this.gridApi, this.gridColumnApi, this.gridMode);
		this.gridOptions.onColumnVisible = (event: ColumnEvent) => this.onColumnVisibilityChanged(event);
		this.restoreGridState({
			groupColumns: this.columnController.selectedGroupColumns,
			columns: this.columnController.columnState,
			filters: this.filterController?.selectedFilters,
			sortModel: this.sortController?.sortModel
		});
		this.updateColumnDefs();
		this.gridReady.next(this);

		tableModalWindowFix();
	}

	public onFirstDataRendered(params) {
		this._firstDataRendered$.next(true);
		setTimeout(() => {
			this.adjustGridSize();
		});
	}

	public adjustGridSize() {
		this.setColumnWidth();
	}

	public getGridState(): IInternalStateSG {
		return {
			groupColumns: this.gridColumnApi.getRowGroupColumns().map((x) => x.getColId()),
			columns: this.gridColumnApi.getColumnState(),
			filters: this.gridApi.getFilterModel(),
			sortModel: getGridSortModel(this.gridColumnApi)
		};
	}

	public restoreGridState(state: IInternalStateSG) {
		this.columnController.updateGroupColumns(state.groupColumns);
		this.columnController.updateColumnState(state.columns);
		this.filterController?.updateFilters(state.filters);
		this.sortController?.updateSortModel(state.sortModel);
	}

	public restoreDefaultColumnsVisibility(gridState: IInternalStateSG) {
		gridState.columns
			.filter(column => this.columnMetaData.some(colMeta => colMeta.columnName === column.colId && colMeta.visible))
			.forEach(column => column.hide = false);
	}

	public updateColumnDefs() {
		const state = this.getGridState();
		const hasVisibleColumns = !!this.columnController.getVisibleColumns().length;
		// if all columns were hidden by user - restore all column visibility on refresh
		if (state.columns && !hasVisibleColumns) {
			this.restoreDefaultColumnsVisibility(state);
		}
		this.restoreGridState(state);
	}

	public get emptyMessage(): string {
		return this.options && this.options.emptyMessage;
	}

	private isExcludeFromExport(colDef: ColDef): boolean {
		return !!(colDef.refData && colDef.refData.excludeFromExport);
	}

	public exportData(): void {
		if (!this.gridApi && !this.gridColumnApi) { return; }
		let fileName = "Dashworks export";
		if (!!this.options.exportFileName) {
			fileName = this.options.exportFileName.replace(/[^0-9a-z_ \-\(\)]/gi, "");
		}
		fileName += ` - ` + `${moment().format("YYYYMMDD")}`;

		let columnKeys = this.gridColumnApi.getAllColumns()
			.filter((column: Column) => column.isVisible() && !this.isExcludeFromExport(column.getColDef()))
			.map((column: Column) => column.getColId());

		columnKeys = intersection(columnKeys, this.columnMetaData.map((columnMeta: IColumnMetaSG) => columnMeta.columnName));
		const params: CsvExportParams = {
			allColumns: false,
			fileName,
			columnKeys,
			processCellCallback: (cellForExportParams: ProcessCellForExportParams): string => {
				return getValueByType(cellForExportParams.column.getColDef(), cellForExportParams.value);
			}
		};
		this.gridApi.exportDataAsCsv(params);
	}

	private setColumnWidth() {
		if (this.gridColumnApi) {
			this.gridColumnApi.autoSizeAllColumns();
		}
	}

	@HostListener("window:resize", ["$event"])
	public onResize() {
		this.adjustGridSize();
	}

	@HostListener("document:click", ["$event"])
	onClick(event) {
		if (checkClass(event.target, "ag-header-cell-menu-button") || checkClass(event.target, "mat-option")) {
			aggridMenuFix();
		}
		if (checkClass(event.target, "mat-checkbox") || checkClass(event.target, "ag-menu-option")) {
			aggridMenuFix();
		}
	}

	public ngOnDestroy() {
		this.destroy$.next();
		this.destroy$.complete();
		if (document.getElementById("content")) {
			document.getElementById("content").style.overflow = "auto";
		}
	}
}
