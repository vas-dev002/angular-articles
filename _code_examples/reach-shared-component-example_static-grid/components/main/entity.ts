import { Type } from "@angular/core";
import {
	GetContextMenuItems,
	GetMainMenuItems,
	GridApi,
	RowNode,
	CellValueChangedEvent,
	ICellEditorParams
} from "@ag-grid-community/core";
import { ColumnState } from "@ag-grid-community/core/dist/cjs/columnController/columnController";
import { ICellRendererAngularComp, ICellEditorAngularComp } from "@ag-grid-community/angular";
import { IEditorTextValidator } from "../columns/editors/textEditor/text-editor.entity";
import { IColDefAbstractFactory } from "../columns/types/base/base-abstract-factory/models";

export interface IColumnMetaSG {
	columnName: string;
	lockVisible: boolean;
	visible: boolean;
	cssClass: string;
	headerName: string;
	controller: Type<IColDefAbstractFactory>;
	filterType: FilterTypeSG;
	isObjectKey: boolean;
	/** The temporary field until static grid has a connection with old version*/
	metaFromApi?: Object;
	/** The column name from data with tooltip*/
	tooltipField?: string;
	/** The column name from data with status of availability*/
	disabledField?: string;
}

export enum FilterTypeSG {
	Multiselect = "Multiselect",
	Number = "Number",
	String = "String",
	Date = "Date",
	DateTime = "DateTime",
	None = "None"
}

export interface IServiceColumn {
	serviceComponent: Type<ICellRendererAngularComp>;
	serviceParams: Object;
	index: number;
}

/** Configuration of Static Grid*/
export interface IOptionsSG {
	/** The message will be shown if data is not provided*/
	emptyMessage?: string;
	disableGroupBy?: boolean;
	editorsByKey?: { [columnKey: string]: IEditorOptions<{}> };
	serviceColumns?: IServiceColumn[];
	enableCheckBox?: boolean;
	suppressHeaderCheckBox?: boolean;
	dataKey?: DataKeyTypeSG;
	/** The object will be provided as a part of params for each column controller*/
	columnControllerOptions?: Object;
	exportFileName?: string;
	/** The options will be copied into agGrid options (GridOptions type)*/
	agGridOptions?: IAgGridOptions;
	styles?: Object;
	enableRowDragAndDrop?: boolean;
	suppressFilter?: boolean;
	suppressSortable?: boolean;
	suppressFloatingFilter?: boolean;
}

export interface IOptionsDynamicSG {
	dataUrl: string;
	agGridOptions?: IDynamicAgGridOptions;
}

export interface IShowingSelectionOptions {
	showCheckBoxes?: boolean;
	suppressHeaderCheckBox?: boolean;
}

export interface IDynamicAgGridOptions {
	cacheBlockSize?: number;
	maxBlocksInCache?: number;
	blockLoadDebounceMillis?: number;
}

export interface IAgGridOptions {
	getContextMenuItems?: GetContextMenuItems;
	getMainMenuItems?: GetMainMenuItems;
	localeTextFunc?: (key: string, defaultValue: string) => string;
	/** Define conditions for styles. Do not forget about ViewEncapsulation in Angular Components*/
	getRowClass?: (params: IGetRowClassParams<any>) => (string | string[]);
}

export interface IGetRowClassParams<TDataRow> {
	api: GridApi;
	data: TDataRow;
	node: RowNode;
	rowIndex: number;
}

export interface IGetCellClassParams<TDataRow, TCellValue> {
	value: TCellValue;
	api: GridApi;
	data: TDataRow;
	node: RowNode;
	rowIndex: number;
}

export interface IEditorOptions<Tvalidator> {
	editor: Type<ICellEditorAngularComp>;
	onChanged: (params: CellValueChangedEvent) => void;
	validator?: Tvalidator;
}

export interface ICellEditorParamsSG<Tvalidator> extends ICellEditorParams {
	validator?: Tvalidator;
}

export interface IEditorParams {
	meta: IColumnMetaSG;
	validator?: IEditorTextValidator;
}

export interface IInternalStateSG {
	groupColumns: string[];
	columns: ColumnState[];
	filters: ISelectedFilters;
	sortModel: ISortModel[];
}
export const defaultDataKeySG = "objectKey";
export type DataKeyTypeSG = string;
export type DataIdTypeSG = string | number;
export interface IDataRowSG {
	[key: string]: any;
}

export interface IFilterModel {
	type: string;
	filter: string;
	filterType: string;
}

export interface ISelectedFilters {
	[key: string]: IFilterModel;
}
export interface ISortModel {
	colId: string;
	sort: string;
}

export interface IDragAndDropModel {
	key: string | number;
	index: number;
}

export interface CellToUpdate {
	type: Type<ICellRendererAngularComp>;
	value: any;
}

export enum GridMode {
	StaticMode,
	DynamicMode,
	KeyValueMode
}

export enum RowModelType {
	ClientSide = "clientSide",
	Infinite = "infinite",
	Viewport = "viewport",
	ServerSide = "serverSide"
}
