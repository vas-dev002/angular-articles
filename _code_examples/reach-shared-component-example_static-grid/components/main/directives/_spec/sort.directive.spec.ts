import { TestBed, ComponentFixture } from "@angular/core/testing";
import { Component, ViewChild } from "@angular/core";
import { StaticGridComponent } from "../../comp";
import { ColDef, GridOptions, SortChangedEvent } from "@ag-grid-community/core";
import { AgGridModule } from "@ag-grid-community/angular";
import { TranslateService, TranslateModule } from "@ngx-translate/core";
import { SortControllerSGDirective } from "../sort.directive";
import { GridMode } from "../../entity";

declare const expect: jest.Expect;

@Component({
	template: `
		<static-grid sortControllerSG="" [sortModel]="sortModel"></static-grid>
	`
})
class TestComponent {

	@ViewChild(SortControllerSGDirective, { static: true })
	public sortController: SortControllerSGDirective;

	columnApi: any = {
		setRowGroupColumns() { },
		setColumnState() { },
		getAllDisplayedColumns() {
			return [] as ColDef[];
		},
		getColumnState() {
			return [];
		}
	};

	gridApi: any = {
		getFilterModel() {
			return [];
		},
		setFilterModel() { },
		getSortModel() { },
		setSortModel() { }
	};


	gridOptions = {
		onSortChanged() { }
	} as GridOptions;

	mode = GridMode.StaticMode;
}

describe("Static Grid Sort Controller", () => {
	let component: TestComponent;
	let fixture: ComponentFixture<TestComponent>;
	beforeEach(() => {
		TestBed.configureTestingModule({
			imports: [TranslateModule.forRoot(), AgGridModule.withComponents([])],
			providers: [TranslateService],
			declarations: [
				StaticGridComponent,
				TestComponent,
				SortControllerSGDirective
			]
		});
	});

	beforeEach(() => {
		fixture = TestBed.createComponent(TestComponent);
		component = fixture.componentInstance;
		component.sortController.init(
			component.gridOptions,
			component.gridApi,
			component.columnApi,
			component.mode
		);
		fixture.detectChanges();
	});

	it("should compile", () => {
		expect(component).toBeTruthy();
	});

	it("should return sortModel", () => {
		component.sortController.sortModel = [];
		const sortModel = component.sortController.sortModel;
		expect(sortModel).toMatchSnapshot();
	});

	it("should be called setSortModel", () => {
		const spyOnGetColumnState = jest.spyOn(component.columnApi, "getColumnState");
		component.sortController.sortModel = [];
		expect(spyOnGetColumnState).toHaveBeenCalled();
	});

	it("should call change event if sort model is changed", () => {
		const spyOnSetSortModel = jest.spyOn(component.sortController.sortModelChange, "next");
		component.gridOptions.onSortChanged({} as SortChangedEvent);
		expect(spyOnSetSortModel).toHaveBeenCalled();
	});
});
