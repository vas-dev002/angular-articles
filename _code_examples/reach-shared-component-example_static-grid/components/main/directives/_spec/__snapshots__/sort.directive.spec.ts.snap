// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Static Grid Sort Controller should return sortModel 1`] = `Array []`;
